# **Pandas Assignment**
 
### **Name:** Tarani 
 **A - Output Prediction |
 B - Code Correction |
 C - Edge Cases |
D - Scenario-Based Questions**


## 1. Installing Pandas

Pandas can be installed using `pip`. The following command installs the Pandas library in the Python environment.

In [1]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


## 2. Importing Pandas

After installing Pandas, we import the library into the Python program so that we can use its functions and DataFrame operations.

In [2]:
import pandas as pd

# **Section A — Output Prediction**

## Q1. Array operations and vectorization

**Predicted output**
```
[20 40 60]
0    20
1    40
2    60
dtype: int64
```

In [3]:
import numpy as np
import pandas as pd

# Create a NumPy array
arr = np.array([10, 20, 30])

# Create a Pandas Series
s = pd.Series([10, 20, 30])

# Multiply each element of the NumPy array by 2
print(arr * 2)

# Multiply each element of the Pandas Series by 2
print(s * 2)

[20 40 60]
0    20
1    40
2    60
dtype: int64


**Explanation:** Both multiply every element by 2 at once (vectorization, no loop needed); the array returns bare values, while the Series returns values together with its index and dtype.

- **Similarity:** Both NumPy arrays and Pandas Series support vectorized operations such as multiplying every element by 2.

- **Difference:** A NumPy array is mainly used for numerical computation, while a Pandas Series has an index and is designed for labeled data.

## Q2. Boolean Filtering

**Predicted output** (rows where Marks >= 65: B, C, D; original index labels 1, 2, 3 are kept)
```
  Student  Marks
1       B     80
2       C     65
3       D     90
```

In [4]:
import pandas as pd

# Create a DataFrame with Student names and their Marks
df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

# Filter the rows where Marks are greater than or equal to 65
result = df[df['Marks'] >= 65]

# Display only the Student and Marks columns
print(result[['Student', 'Marks']])
print(result[['Student', 'Marks']].shape)

  Student  Marks
1       B     80
2       C     65
3       D     90
(3, 2)


**Explanation:** `df['Marks'] >= 65` builds a True/False mask, and indexing with it keeps only the True rows (B, C, D), so the result has **3 rows and 2 columns**, shape `(3, 2)`.

## Q3. Descriptive statistics

**Predicted output**
```
30.0
50
30.0
```

In [5]:
import pandas as pd

# Create a DataFrame with a Marks column
df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

# Calculate and print the average (mean) of Marks
print("Mean:",df['Marks'].mean())

# Find and print the highest mark
print("Highest mark:", df['Marks'].max())

# Calculate and print the middle value (median) of Marks
print("Median:",df['Marks'].median())

Mean: 30.0
Highest mark: 50
Median: 30.0


**Explanation:** The mean is the sum divided by the count (150/5), the max is the largest value, and the median is the middle value of the sorted data.

- **Median vs mean:** the mean uses every value, so extreme values pull it up or down; the median depends only on the middle position, so it resists outliers. Here the data is evenly spaced, so both equal 30.0.

## Q4. Group-wise Analysis

**Predicted output**
```
Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64
```

In [6]:
import pandas as pd

# Create a DataFrame with Department and Salary columns
df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

# Group the data by Department and calculate the average Salary for each department
print(df.groupby('Department')['Salary'].mean())

Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64


**Explanation:** `groupby()` splits the rows by department, computes the mean of `Salary` inside each group (IT: (40000+60000)/2 = 50000, HR: (30000+50000)/2 = 40000), then combines the results into a Series indexed by department (sorted alphabetically).

## Q5. Pandas Correlation - Positive

**Predicted output:** `1.0`

In [7]:
import pandas as pd

# Create a DataFrame with Hours and Marks columns
df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

# Calculate and print the correlation between Hours and Marks
print("Correlation between Hours and Marks:", df['Hours'].corr(df['Marks']))

Correlation between Hours and Marks: 1.0


**Explanation:** Marks = 20 × Hours exactly, a perfect straight line with a rising slope, so Pearson's r = **+1.0**: a perfect, strong, positive relationship (more study hours means higher marks).

## Q6. Pandas Correlation - Negative

**Predicted output:** `-1.0`

In [8]:
import pandas as pd

# Create a DataFrame with Price and Demand columns
df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

# Calculate and print the correlation between Price and Demand
print("Correlation between Price and Demand:", df['Price'].corr(df['Demand']))

Correlation between Price and Demand: -1.0


**Explanation:** Demand falls by exactly 20 for every 10 increase in Price (a perfect straight line with a falling slope), so r = **-1.0**; the negative sign means the variables move in opposite directions (higher price, lower demand).

# Section B — Code Correction

## Q1. JSON and CSV

**Issue:** `header=None` tells Pandas the file has no header row, so the header line is treated as data (and columns get numbers '0', '1', '2'). Since the objective is to have the header row used as column names, remove `header=None` (the default `header='infer'` uses the first row).

In [9]:
import pandas as pd

# Read the CSV file and use the first row as column names
df = pd.read_csv("students.csv")

# Display the DataFrame
print(df)

    ID     Name        Date  Marks
0  101  Asha, R  2026-09-01   80.0
1  102     Ravi  01/09/2026    NaN
2  103     Neha  2026-09-03    NaN


**Explanation:** `pd.read_csv()` reads the CSV file into a DataFrame. By default, Pandas uses the first row as the column names, so removing `header=None` allows the header to be read correctly.

## Q2. Data Analysis

**Issue:** `sort_values('Salary')` sorts the Salary column in ascending order by default, so the smallest salaries appear first. As a result, `head(5)` returns the five lowest salaries. Since the objective is to display the five highest salaries, use `ascending=False` to sort from highest to lowest.

In [10]:
import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary', ascending=False).head(5))

  Employee  Salary
5        F   70000
3        D   60000
1        B   50000
4        E   45000
0        A   30000


**Explanation:** Adding `ascending=False` (or using `nlargest(5, 'Salary')`) puts the highest salaries first, so `head(5)` returns the top five.

## Q3. Cleaning Data

**Issue:** `fillna()` returns a *new* Series and does not change the DataFrame; the result was never assigned back, so `df` still has the missing value. Assign the result back to the column.

In [11]:
import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

# Incorrect: result is thrown away
df['Marks'].fillna(df['Marks'].mean())
print(df)

# Corrected
df['Marks'] = df['Marks'].fillna(df['Marks'].mean())
print(df)

   Marks
0   70.0
1    NaN
2   90.0
   Marks
0   70.0
1   80.0
2   90.0


**Explanation:** Assigning the result back (`df['Marks'] = ...`) stores the filled values; the mean of the non-missing marks is (70 + 90) / 2 = 80.

## Q4. Correlation

**Issue:** `DataFrame.corr()` computes a correlation *matrix* of all columns and its first arguments are `method` and `min_periods`, not two column names, so `df.corr('Hours', 'Marks')` raises an error. To correlate two columns, call `.corr()` on one Series and pass the other Series. (The paper's curly quote `’` is also replaced with a straight quote.)

In [12]:
import pandas as pd # Create a DataFrame with Hours and Marks columns 
df = pd.DataFrame({ 'Hours': [1, 2, 3, 4], 
                   'Marks': [20, 40, 60, 80] }) 
# Calculate the correlation between Hours and Marks 
print("Correlation between Hours and Marks:", df['Hours'].corr(df['Marks']))

Correlation between Hours and Marks: 1.0


**Explanation:** `df['Hours']` selects the Hours column, and `df['Marks']` selects the Marks column. .`corr()` calculates the correlation between these two columns. The output `1.0 `indicates a perfect positive correlation in this dataset.

# Section C — Edge Cases

## Q1. Broadcasting with shapes (3, 1) and (1, 3)

In [13]:
import numpy as np

# Column array
a = np.array([[10], [20], [30]])

# Row array
b = np.array([[1, 2, 3]])

# Add arrays using broadcasting
result = a + b

print(result)

# Result shape
print(result.shape)

print("Shape of a:", a.shape)

print("Shape of b:", b.shape)

[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)
Shape of a: (3, 1)
Shape of b: (1, 3)


1. **Output:** a 3 × 3 grid `[[11, 12, 13], [21, 22, 23], [31, 32, 33]]`, shape `(3, 3)`.
2. **Shapes before addition:** `a` is `(3, 1)` and `b` is `(1, 3)`.
3. **How broadcasting works:** NumPy compares shapes from the last dimension backwards; two dimensions are compatible if they are equal or one of them is 1. A size-1 dimension is virtually stretched (no data is copied) to match the other: `a` is stretched across 3 columns and `b` across 3 rows, both become `(3, 3)`, and then they are added element-wise.
4. **Why it differs from adding two 1-D arrays:** two 1-D arrays of shape `(3,)` are added position by position (`[10,20,30] + [1,2,3] = [11,22,33]`, shape `(3,)`). Here every element of `a` is combined with every element of `b` (an "outer sum"), giving a 2-D grid.

**Part 4 — replacing `b` with `np.array([1, 2])`**

In [14]:
import numpy as np

a = np.array([[10], [20], [30]])
b2 = np.array([1, 2])

print("Shape of b2:", b2.shape)
print(a + b2)
print((a + b2).shape)

# The case that actually fails: a 1-D array of length 3 plus a 1-D array of length 2
try:
    print(np.array([10, 20, 30]) + b2)
except ValueError as e:
    print("ValueError -", e)

Shape of b2: (2,)
[[11 12]
 [21 22]
 [31 32]]
(3, 2)
ValueError - operands could not be broadcast together with shapes (3,) (2,) 


**Explanation:** `b2` has shape `(2,)`, which broadcasts as `(1, 2)`; since `a` is `(3, 1)`, the dimensions 3 vs 1 and 1 vs 2 are both compatible, so it does **not** fail and returns a `(3, 2)` result `[[11, 12], [21, 22], [31, 32]]`. Broadcasting only raises an error when a pair of dimensions is unequal and neither is 1 (e.g. `(3,)` with `(2,)`, as shown above).

## Q2. Parsing a CSV with quoted commas, missing values and mixed date formats

In [15]:
import pandas as pd

# Create a CSV file
with open("students.csv", "w") as f:
    f.write('ID,Name,Date,Marks\n'
            '101,"Asha, R",2026-09-01,80\n'
            '102,Ravi,01/09/2026,\n'
            '103,Neha,2026-09-03,NA\n')

# Read the CSV file
df = pd.read_csv("students.csv")

# Display the data
print(df)

# Display the data types
print(df.dtypes)

    ID     Name        Date  Marks
0  101  Asha, R  2026-09-01   80.0
1  102     Ravi  01/09/2026    NaN
2  103     Neha  2026-09-03    NaN
ID         int64
Name         str
Date         str
Marks    float64
dtype: object


1. **Quoted comma:** `read_csv` treats text inside double quotes as a single field (`quotechar='"'`), so `"Asha, R"` stays one `Name` value with 4 columns intact. If the quotes were missing, the comma would split it into an extra column and shift the row.
2. **Date and Marks issues:**
   - `Date` has two formats (`2026-09-01` and `01/09/2026`), so it is loaded as plain text, not as dates, and sorting or date arithmetic would be wrong. `01/09/2026` is also ambiguous (1 September or 9 January?), so the intended format must be decided (assumed day/month/year below).
   - `Marks` contains a blank and `NA`; both silently become `NaN` by default, which forces the column to `float64` (80.0) and hides the fact that they may mean different things.

In [16]:
import pandas as pd

# Keep "NA" as text (only truly blank cells become NaN) so it can be inspected first
df = pd.read_csv("students.csv", dtype=str, keep_default_na=False, na_values=[""])

# Inspect the "NA" entries before deciding what they mean
print(df[df["Marks"] == "NA"])

# Marks -> numeric; anything that is not a number ("NA", blank) becomes NaN
df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")

# Dates -> datetime: parse each known format separately (assumed dd/mm/yyyy for the slash format)
iso = pd.to_datetime(df["Date"], format="%Y-%m-%d", errors="coerce")
dmy = pd.to_datetime(df["Date"], format="%d/%m/%Y", errors="coerce")
df["Date"] = iso.fillna(dmy)

print(df)
print(df.dtypes)

    ID  Name        Date Marks
2  103  Neha  2026-09-03    NA
    ID     Name       Date  Marks
0  101  Asha, R 2026-09-01   80.0
1  102     Ravi 2026-09-01    NaN
2  103     Neha 2026-09-03    NaN
ID                  str
Name                str
Date     datetime64[us]
Marks           float64
dtype: object


3. **Explanation:** `na_values=[""]` and `pd.to_numeric(..., errors="coerce")` turn missing or non-numeric marks into `NaN`, and parsing each date format explicitly (with `format=` and `fillna`) converts the mixed strings into one `datetime64` column without guessing.
4. **Why investigate `"NA"`:** it may mean "Not Available" (marks not entered), "Not Applicable" or "Absent" (the student did not take the exam), or even be a real value or a typing error; each needs different treatment (a student who was absent is not the same as missing data, and imputing or deleting blindly could bias the results), so its meaning should be confirmed with the data source first.

## Q3. Outliers: mean vs median

In [17]:
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())

# Part 4: record with the highest salary
print(df.nlargest(1, "Salary"))

Mean: 2522500.0
Median: 32500.0
  Employee    Salary
3        D  10000000


1. **Prediction:** Mean = (25000 + 30000 + 35000 + 10000000) / 4 = **2522500.0**; Median = (30000 + 35000) / 2 = **32500.0**.
2. **Effect of the extreme value:** the single value 10,000,000 drags the mean up to about 78 times the median, higher than three of the four salaries by a huge margin, so the mean no longer describes anyone typical.
3. **Better statistic:** the **median (32,500)**, because it depends only on the middle values and is not affected by how large the outlier is; it sits close to the three ordinary salaries.
4. **Code:** `df.nlargest(1, "Salary")` (above) returns Employee D. **Explanation:** the record should be investigated before removal because it could be a data-entry error (e.g. extra zeros) or a genuine value (e.g. a CEO's salary); removing a genuine value would lose real information and distort the dataset, while keeping an error would distort the mean.

# Section D — Scenario-Based Questions

## Q1. Cleaning a dirty student assessment dataset

**Step 1 — Load and inspect the Dataset**

In [18]:
import pandas as pd

# Load the tab-separated CSV file
df = pd.read_csv("student_assessment_dirty.csv", sep="\t")

# Display the first five rows
print(df.head())

print()

# Display dataset information
df.info()

print()

# Display missing values in each column
print(df.isnull().sum())

  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python     NaN
4       S105        Meera         HR     Python  Absent

<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         12 non-null     str  
dtypes: str(5)
memory usage: 1.0 KB

Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           3
dtype: int64


**Explanation:**

- `read_csv()` loads the student assessment CSV file into Pandas.
- `head()` and `info()` help us understand the data and its structure.
- `isnull()`.`sum()` shows the number of missing values in each column.

**Step 2 — Standardize department names**

In [19]:
# Remove spaces and dots and convert department names to uppercase
df["Department"] = (
    df["Department"]
    .str.strip()
    .str.replace(".", "", regex=False)
    .str.upper()
)

# Standardize the department names
df["Department"] = df["Department"].replace({
    "IT": "IT",
    "FINANCE": "Finance",
    "HR": "HR"
})

# Display unique department names
print(df["Department"].unique())

<ArrowStringArray>
['IT', 'HR', 'Finance']
Length: 3, dtype: str


**Explanation:** 
- `strip()`, `replace()`, and `upper()` remove formatting differences.
Values such as `IT`, it, and `I.T`. are converted to `IT`.
Similarly, finance and Finance are standardized to Finance.

**Step 3 — Remove exact duplicate records** (done after standardizing, so `finance` vs `Finance` copies of the same record are also caught)

In [20]:
# Find and display duplicate records
print("Duplicate records:")
print(df[df.duplicated()])

# Count the duplicate records
duplicates_removed = df.duplicated().sum()

# Remove exact duplicate records
df = df.drop_duplicates()

# Display the number of records removed
print("Records removed:", duplicates_removed)

Duplicate records:
  Student_ID Student_Name Department Assessment Marks
8       S102         Ravi         IT     Python    78
Records removed: 1


**Explanation:** 
- `duplicated()` identifies repeated student - assessment records.
- `drop_duplicates()` removes the exact duplicate records.
- The duplicate count is stored so we can report it later.

**Steps 4  — Convert Marks to numeric format**

In [21]:
# Convert Marks to string so text values can be cleaned
df["Marks"] = df["Marks"].astype(str)

# Remove the word "marks" from values such as "88 marks"
df["Marks"] = (
    df["Marks"]
    .str.replace("marks", "", case=False, regex=False)
    .str.strip()
)

# Convert the cleaned values to numbers
# Invalid values such as Absent and NA become NaN
df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")

# Display the Marks column
print(df["Marks"])

0      85.0
1      78.0
2      92.0
3       NaN
4       NaN
5       0.0
6      67.0
7      75.0
9       NaN
10     88.0
11     91.0
12      NaN
13     64.0
14    105.0
Name: Marks, dtype: float64


**Explanation:**
- The text `"marks"` is removed from values such as `"88 marks"`.
- `pd.to_numeric()` converts valid mark values into numbers.
- Values such as `"Absent"` and `"NA"` become NaN instead of causing an error.

**Step 5 — Investigate Missing Marks**

In [22]:
# Display records where marks are missing
print("Missing mark records:")
print(df[df["Marks"].isnull()][
    ["Student_ID", "Student_Name", "Marks"]
])

# Count missing marks
missing_marks = df["Marks"].isnull().sum()

print("Missing marks:", missing_marks)

Missing mark records:
   Student_ID Student_Name  Marks
3        S104        Karan    NaN
4        S105        Meera    NaN
9        S109        Sneha    NaN
12       S112         Aman    NaN
Missing marks: 4


**Explanation:** 
- `isnull()` identifies students whose marks are missing.
- We investigate these records before deciding whether to delete them.
- Deleting missing values without investigation may remove important information or introduce bias.

**Step 6 — Identify Invalid Marks**

In [23]:
# Find marks outside the valid range of 0 to 100
invalid_marks = df[
    df["Marks"].notnull() &
    ~df["Marks"].between(0, 100)
]

# Display invalid marks
print("Invalid marks:")
print(invalid_marks[["Student_ID", "Student_Name", "Marks"]])

# Count invalid marks
print("Invalid marks identified:", len(invalid_marks))

# Replace invalid marks with missing values
df.loc[
    df["Marks"].notnull() &
    ~df["Marks"].between(0, 100),
    "Marks"
] = pd.NA

Invalid marks:
   Student_ID Student_Name  Marks
14       S114        Kabir  105.0
Invalid marks identified: 1


**Explanation:** 
- `between(0, 100)` checks whether marks are within the valid range.
The value `105` is identified as an invalid mark.
- Invalid marks are replaced with `NaN` because they need correction or verification.

**Step 7 - Display Cleaned Data and Summary**

In [24]:
# Display the final cleaned DataFrame
print("Cleaned DataFrame:")
print(df)

print()

# Display the cleaning summary
print("Records removed:", duplicates_removed)
print("Missing marks remaining:", df["Marks"].isnull().sum())
print("Invalid marks identified:", len(invalid_marks))

Cleaned DataFrame:
   Student_ID Student_Name Department Assessment  Marks
0        S101         Asha         IT     Python   85.0
1        S102         Ravi         IT     Python   78.0
2        S103         Neha         IT     Python   92.0
3        S104        Karan         HR     Python    NaN
4        S105        Meera         HR     Python    NaN
5        S106        Arjun    Finance     Python    0.0
6        S107        Priya    Finance     Python   67.0
7        S108        Rahul         IT     Python   75.0
9        S109        Sneha         IT     Python    NaN
10       S110        Varun         HR     Python   88.0
11       S111         Diya    Finance     Python   91.0
12       S112         Aman         IT     Python    NaN
13       S113        Nisha         HR     Python   64.0
14       S114        Kabir    Finance     Python    NaN

Records removed: 1
Missing marks remaining: 5
Invalid marks identified: 1


**Explanation:** 
- The cleaned DataFrame is displayed after all cleaning operations.
- We report how many duplicate records were removed.
- We also report the remaining missing marks and invalid marks identified.